src/main/java/uk/nhs/hee/trainee/details/service/NtnGenerator.java

```diff
@@ -0,0 +1,303 @@
+   *
+   * @param programmeMembership The programme membership to get the specialty string for.
+   * @return The concatenated specialty string.
+   */
+  private String getSpecialtyConcat(ProgrammeMembership programmeMembership) {
+    log.info("Calculating specialty concat.");
+    List<Curriculum> curricula = programmeMembership.getCurricula();
+    List<Curriculum> sortedCurricula = filterAndSortCurricula(curricula);
+
+    StringBuilder sb = new StringBuilder();
+
+    for (ListIterator<Curriculum> curriculaIterator = sortedCurricula.listIterator();
+        curriculaIterator.hasNext(); ) {
+      int index = curriculaIterator.nextIndex();
+      Curriculum curriculum = curriculaIterator.next();
+      String specialtyCode = curriculum.getCurriculumSpecialtyCode();
+
+      if (index > 0) {
+        if (curriculum.getCurriculumSubType().equals("SUB_SPECIALTY")) {
+          log.info("Appending sub-specialty '{}'.", specialtyCode);
+          sb.append(".");
+        } else {
+          log.info("Appending specialty '{}'.", specialtyCode);
+          sb.append("-");
+        }
+      } else {
+        log.info("Using '{}' as first specialty.", specialtyCode);
+      }
+
+      sb.append(specialtyCode);
+
+      if (index == 0 && Objects.equals(curriculum.getCurriculumName(), "AFT")) {
+        sb.append("-FND");
+        break;
+      }
+    }
+
+    log.info("Calculated specialty concat: '{}'.", sb);
+    return sb.toString();
+  }
+
+  /**
+   * Get the GMC/GDC reference number for the given personal details.
+   *
+   * @param personalDetails The personal details to get the reference number for.
+   * @return The GMC/GDC number, depending on which is valid.
+   */
+  private String getReferenceNumber(PersonalDetails personalDetails) {
+    String gmcNumber = personalDetails.getGmcNumber();
+    return gmcNumber.matches("\\d{7}") ? gmcNumber : personalDetails.getGdcNumber();
+  }
+
+  /**
+   * Get the NTN suffix for the given programme membership.
+   *
+   * @param programmeMembership The programme membership to get the suffix for.
+   * @return The calculated suffix for the programme membership's NTN.
+   */
+  private String getSuffix(ProgrammeMembership programmeMembership) {
+    log.info("Calculating suffix.");
+    String trainingPathway = programmeMembership.getTrainingPathway();
+    log.info("Using training pathway '{}' to calculate suffix.", trainingPathway);
+
+    String suffix = switch (trainingPathway) {
+      case "CCT" -> "C";
+      case "CESR" -> "CP";
+      default -> {
+        List<Curriculum> curricula = programmeMembership.getCurricula();
+        List<Curriculum> sortedCurricula = filterAndSortCurricula(curricula);
+        String firstSpecialtyCode = sortedCurricula.get(0).getCurriculumSpecialtyCode();
+        log.info("Using specialty code '{}' to calculate suffix.", trainingPathway);
+
+        yield Objects.equals(firstSpecialtyCode, "ACA") ? "C" : "D";
```

Comment:
Was it ok to collapse the SQL logic here? 
```
when
--(
prog.ProgrammeName like '%Core%'
or
prog.ProgrammeName like '%ACCS%'
or
prog.ProgrammeName like '%Acute%Care%Common%Stem%'
or
prog.ProgrammeName like '%Combined%Infection%Training%'
or
Prog.ProgrammeName like 'Internal%Medicine%Training%'
or
rcm1.TrainingPathway = 'N/A'
--)
--and
--rcm1.FirstSpecialtyNumber <> 'ACA'
Then 'D'
```

Reply:
I did confirm with James, he couldn't see a reason not to (and my (now removed) unit tests all stayed green too).
The only time the programme name conditions would be relevant were if the TrainingPathway had an invalid/null value, but other than 4 programme memberships (0.00001%) it's all standardised and `N/A` is properly used as the default.